docker/package/util/build.py

```diff
@@ -0,0 +1,208 @@
+
+    if (args.sources_dir or args.launchpad_sources) and target_os != "ubuntu":
+        raise Exception("Sources-related options are only supported for Ubuntu.")
+
+    if args.sources_dir and args.launchpad_sources:
+        raise Exception(
+            "--sources-dir and --launchpad-sources options are mutually exclusive."
+        )
+
+    # copr build infrastructure uses latest stable fedora and `mock` for builds
+    # so we should also keep that way
+    # for ubuntu builds, since we lack `pbuilder` for now,
+    # packages should be built in respective containers for better reproducibility
+    images = ["37"] if target_os == "fedora" else distributions
+
+    packages_to_build = get_packages_to_build(args.packages)
+
+    if not args.build_sapling_package:
+        packages_to_build.pop("tezos-sapling-params", None)
+
+    artifacts = []
+
+    for image in images:
+
+        # prebuild docker image before using containers
+        check_call(
+            f"""
+        {virtualisation_engine}
+        build -t tezos-{target_os}-{image}
+        -f docker/package/Dockerfile-{target_os} --build-arg OCTEZ_VERSION={octez_version} --build-arg dist={image} .
+        """
+        )
+
+        distros = [image] if target_os == "ubuntu" else distributions
+        cmd_args = " ".join(
+            [
+                f"--os {target_os}",
+                f"--binaries-dir {binaries_dir_name}" if binaries_dir_name else "",
+                f"--sources-dir {sources_dir_name}" if sources_dir_name else "",
+                f"--type {args.type}",
+                f"--distributions {' '.join(distros)}",
+                f"--launchpad-sources" if args.launchpad_sources else "",
+                f"--packages {' '.join(packages_to_build.keys())}",
+            ]
+        )
+
+        container_create_args = (
+            "--cap-add SYS_ADMIN"
+            if target_os == "fedora" and args.distributions
+            else ""
+        )
+
+        container_id = get_proc_output(
+            f"""
+        {virtualisation_engine}
+        create {" ".join(["-v " + v for v in docker_volumes])}
+        {container_create_args}
+        --env OCTEZ_VERSION={octez_version}
+        --env OPAMSOLVERTIMEOUT=900
+        -t tezos-{target_os}-{image} {cmd_args}
+        """
+        ).stdout.strip()
+
+        exit_code = call(f"{virtualisation_engine} start -a {container_id}")
+
+        os.makedirs(args.output_dir, exist_ok=True)
+
+        call(
+            f"""
+        {virtualisation_engine} cp
+        {container_id}:/tezos-packaging/docker/{container_output_dir}/. {args.output_dir}
+        """
+        )
+
+        call(f"{virtualisation_engine} rm -v {container_id}")
+
+        if exit_code:
+            print("Unrecoverable error occured.")
+            sys.exit(exit_code)
+
+        # the same source archive has to be reused for an ubuntu package on different distros
+        if sources_dir_name is None and target_os == "ubuntu":
```

Comment:
There's quite a bit of differences between the `ubuntu` and `fedora` builds at this point, I see a few `if`s that handle the two differently.

I'm thinking that it would be better to split the builds in two different python source files, both of which import a third file (non-executable) with the common code.

How does that sound to you?

Reply:
@krendelhoff2 I see the new changes, but AFAICT there is still a common `build.py` file.

Have you considered splitting this?

Reply:
@krendelhoff2 I looked at the new changes and AFAICT there's still a common `build.py` for both (and the same goes for signing and uploading).

Have you considered my suggestion to split these?

To be clear: I see that you've split some functions to only do one thing and reduced the condition checks already, but I still think that there's some use in having these in separate file to avoid the argument parsing and make these files even more single-purpose.